## 1.3 Haskell Type Declarations

Haskell has a concise way to indicate that `divides` consumes an integer, then
another integer, and produces a truth value (called `Bool` in Haskell). Integers and truth values are examples of *types*. See Section 2.1 for more on the type `Bool`.
Section 1.6 gives more information about types in general. Arbitrary precision
integers in Haskell have type `Integer`. The following line gives a so-called type
declaration for the `divides` function.

```haskell
divides :: Integer -> Integer -> Bool
```

`Integer -> Integer -> Bool` is short for `Integer -> Integer -> Bool)`. A type of the form `a -> b` classifies a procedure that takes an argument of type `a`
to produce a result of type `b`. Thus, `divides` takes an argument of type `Integer` and produces a result of type `Integer -> Bool`, i.e., a procedure that takes an argument of type `Integer`, and produces a result of type `Bool`.

The full code for `divides`, including the type declaration, looks like this:


In [13]:
divides :: Integer -> Integer -> Bool
divides d n = rem n d == 0

If `d` is an expression of type `Integer`, then `divides d` is an expression of type
`Integer -> Bool`. The shorthand that we will use for

> `d` is an *expression of type* `Integer`

is `d :: Integer`.

### Exercise 1.6

Can you gather from the definition of `divides` what the type declaration for `rem` would look like?

### Exercise 1.7

IHaskell and GHCI have a command for checking the types of expressions. Can you explain the following:

In [14]:
:t divides 5

divides 5 :: Integer -> Bool

In [15]:
:t divides 5 7

divides 5 7 :: Bool

The expression `divides 5 :: Integer -> Bool` is called a *type judgment*. Type judgments in Haskell have the form `expression :: type`.

In Haskell it is not strictly necessary to give explicit type declarations. For instance, the definition of divides works quite well without the type declaration, since the system can infer the type from the definition. However, it is good programming practice to give explicit type declarations even when this is not strictly necessary. These type declarations are an aid to understanding, and they greatly improve the digestibility of functional programs for human readers. A further advantage of the explicit type declarations is that they facilitate detection of programming mistakes on the basis of type errors generated by the interpreter. You will find that many programming errors already come to light when your program gets loaded. The fact that your program is well typed does not entail that it is correct, of course, but many incorrect programs do have typing mistakes.

The fill code for `ld`, including the type declaration, looks like this:

In [16]:
{- HLINT ignore "Eta reduce" -}
ld :: Integer -> Integer
ld n = ldf 2 n

ldf :: Integer -> Integer -> Integer
ldf k n | divides k n = k
        | k^2 > n     = n
        | otherwise   = ldf (k+1) n

The first line of the code states that the operation `ldf` takes two integers and pro-
duces an integer.

The full code for `prime0`, including the type declaration, runs like this:

In [17]:
prime0 :: Integer -> Bool
prime0 n | n < 1 = error "not a positive integer"
         | n == 1 = False
         | otherwise = ld n == n

The first line of the code declares that the operation `prime0` takes an integer and
produces (or *returns*, as programmers like to say) a Boolean (truth value).

In programming generally, it is useful to keep close track of the nature of the objects that are being represented. This is because representations have to be stored in computer memory, and one has to know how much space to allocate for this storage. Still, there is no need to always specify the nature of each data-type explicitly. It turns out that much information about the nature of an object can be inferred from how the object is handled in a particular program, or in other words, from the operations that are performed on that object.

Take again the definition of `divides`. It is clear from the definition that an operation is defined with two arguments, both of which are of a type for which `rem` is defined, and with a result of type `Bool` (for `rem n d == 0` is a statement that can turn out true or false). If we check the type of the built-in procedure rem we get:

In [18]:
:t rem

rem :: forall a. Integral a => a -> a -> a

In this particular case, the type judgment gives a *type scheme* rather than a type. It
means: if a is a type of class `Integral`, then `rem` is of type `a -> a -> a`. Here
`a` is used as a variable ranging over types.

In Haskell, `Integral` is the class (see Section 4.2) consisting of the two types for
integer numbers, `Int` and `Integer`. The difference between `Int` and `Integer`
is that objects of type `Int` have fixed precision, objects of type `Integer` have
arbitrary precision.

The type of `divides` can now be inferred from the definition. This is what we get
when we load the definition of divides without the type declaration:

In [19]:
divides d n = rem n d == 0
:t divides

divides :: forall {a}. Integral a => a -> a -> Bool

## 1.4 Identifiers in Haskell

In Haskell, there are two kinds of identifiers:

* Variable identifiers are used to name functions. They have to start with a lower-case letter. E.g., `map`, `max`, `fct2list`, `fctToList`, `fct_to_list`.
* Constructor identifiers are used to name types. They have to start with an upper-case letter. Examples are `True`, `False`.

Functions are operations on data-structures, constructors are the building blocks of the data structures themselves (trees, lists, Booleans, and so on).

Names of functions always start with lowe-case letters, and may contain both upper- and lower-case letters, but also digits, underscores and the prime symbol `'`. The following *reserved keywords* have special meanings and cannot be used to
name functions.

> `case` `class` `data` `default` `deriving` `do` `else` `if` `import` `in` `infix` `infixl` `infixr` `instance` `let` `module` `newtype` `of` `then` `type` `where` `forall` `error`

The use of these keywords will be explained as we encounter them. `_` at the beginning of a word is treated as a lower-case character. The underscore character `_` all by itself is a reserved word for the wild card pattern that matches anything (**TODO** page 141).

**TODO** Do we need to care about `forall` keyword if not using HUGS?

## 1.5 Playing the Haskell Game

This section consists of a number of further examples and exercises to get you acquainted with the programming language of this notebook.

**TODO** Furhter clarity how to use these examples?

In the next example, we use `Int` for the type of fixed precision integers, and `[Int]` for lists of fixed precision integers.

### Example 1.8 

Here is a function that gives the minimum of a list of integers e.g. [1,2,3,4]:


In [20]:
mnmInt :: [Int] -> Int
mnmInt [] = error "empty list"
mnmInt [x] = x
mnmInt (x:xs) = min x (mnmInt xs)

This uses the predefined function `min` for the minimum of two integers. It also
uses pattern matching for lists. The list pattern `[]` matches only the empty list,
the list pattern `[x]` matches any singleton list, the list pattern `(x:xs)` matches any
non-empty list. A further subtlety is that pattern matching in Haskell is sensitive
to order. If the pattern `[x]` is found before `(x:xs)` then `(x:xs)` matches any
non-empty list that is not a unit list. See Section 4.6 for more information on list
pattern matching.

It is common Haskell practice to refer to non-empty lists as `x:xs`, `y:ys`, and so
on, as a useful reminder of the facts that `x` is an element of a list of `x`'s and that `xs` is a list.

Here is a home-made version of `min`

In [21]:
min' :: Int -> Int -> Int
min' x y | x <= y    = x
         | otherwise = y

You will have guessed that `<=` is Haskell code for $\leq$.

Objects of type `Int` are fixed precision integers. Their range can be found with:

In [22]:
(minBound :: Int)

-9223372036854775808

In [23]:
(maxBound :: Int)

9223372036854775807

Since $9223372036854775807 = 2^{63}-1$ we can conclude that the IHaskell implementation uses eight bytes (64 bits) to represent objects of this type. Integer is for arbitrary precision integers: the storage space that gets allocated for Integer objects depends on the size of the object.

### Exercise 1.9 

Define a function that gives the maximum of a list of integers. Use the predefined function `max`.

### Conversion from Prefix to Infix in Haskell

A function can be converted to an
infix operator by putting its name in back quotes, like this:

In [24]:
max 4 5

5

In [25]:
4 `max` 5

5

Conversely, an infix operator is converted to prefix by putting the operator in round
brackets (p. 21).

### Exercise 1.10

Define a function `removeFst` that removes the first occurrence of an integer $m$ from a list of integers. If $m$ does not occur in the list, the list remains unchanged.



In [26]:
removeFst = undefined

### Exercise 1.11

We define a function that sorts a list of integers in order of increasing size, by means of the following algorithm:

* an empty list is already sorted.
* if an list is non-empty, we put its minimum in front of the result of sorting the list that returns from removing its minimum.

This is implemented as follows:

In [27]:
{- HLINT ignore "Redundant bracket" -}
srtInts :: [Int] -> [Int]
srtInts [] = []
srtInts xs = m : (srtInts (removeFst m xs)) where m = mnmInt xs

Here `removeFst` is the function you defined in Exercise 1.10. Note that the second
clause is invoked when the first one does not apply, i.e., when the argument of
`srtInts` is not empty. This ensures that `mnmInt xs` never gives rise to an error.
Note the use of a `where` construction for the local definition of an auxiliary function.

**Remark**

Haskell has two ways to locally define auxiliary functions, `where` and
`let` constructions. The where construction is illustrated in Example 1.11. This
can also expressed with let, as follows:

In [28]:
{- HLINT ignore "Redundant bracket" -}
srtInts' :: [Int] -> [Int]
srtInts' [] = []
srtInts' xs = let m = mnmInt xs 
              in  m : (srtInts' (removeFst m xs))

The `let` construction uses the reserved keywords `let` and `in`.

### Example 1.12

Here is a function that calculates the average of a list of integers. The average of $m$ and $n$ is given by $\frac{m+n} 2$, the average of a list of $k$ integers $n_1, \ldots , n_k$ is given by $\frac {n_1 + \ldots + n_k} k$. In general, averages are fractions, so the result type of average should not be `Int` but the Haskell data-type for floating point numbers, which is `Float`. There are predefined functions sum for the sum of a list of integers, and `length` for the length of a list. The Haskell operation for division `/` expects arguments of type `Float` (or more precisely, of `Fractional` type, and `Float` is such a type), so we need a conversion function for converting `Ints` into `Floats`. This is done by `fromIntegral `. The function average can now be written as

In [29]:
average :: [Int] -> Float
average [] = error "empty list"
average xs = fromIntegral  (sum xs) / fromIntegral  (length xs)

Again, it is instructive to write our own homemade versions of sum and length.

Here they are:

In [30]:
{- HLINT ignore "Use foldr" -}
sum' :: [Int] -> Int
sum' [] = 0
sum' (x:xs) = x + sum' xs

In [31]:
{- HLINT ignore "Use foldr" -}
length' :: [a] -> Int
length' [] = 0
length' (x:xs) = 1 + length' xs

Note that the type declaration for `length’` contains a variable `a`. This variable
ranges over all types, so `[a]` is the type of a list of objects of an arbitrary type `a`.
We say that `[a]` is a *type scheme* rather than a type. This way, we can use the same
function `length'` for computing the length of a list of integers, the length of a list
of characters, the length of a list of strings (lists of characters), and so on.

The type `[Char]` is abbreviated as `String`. Examples of characters are `'a'`, `'b'`
(note the single quotes) examples of strings are `"Russell"` and `"Cantor"` (note
the double quotes). In fact, `"Russell"` can be seen as an abbreviation of the list:

In [32]:
[ 'R', 'u', 's', 's', 'e', 'l', 'l' ]

"Russell"

### Exercise 1.13

Write a function count for counting the number of occurrences of
a character in a string. In Haskell, a character is an object of type `Char`, and a string an object of type `String`, so the type declaration should run: `count :: Char -> String -> Int`.

### Exercise 1.14 

A function for transforming strings into strings is of type `String -> String`. Write a function blowup that converts a string

$$ a_1 a_2 a_3 \ldots $$

to

$$ a_1 a_2 a_2 a_3 a_3 a_3 \ldots $$

`blowup "bang!"` should yield `"baannngggg!!!!!"`. (Hint: use `++` for string concatenation.)

### Exercise 1.15

Write a function `srtString :: [String] -> [String]` that sorts a list of strings in alphabetical order.

### Exercise 1.16

Suppose we want to check whether a string `str1` is a prefix of a string `str2`. Then the answer to the question `prefix str1 str2` should be either yes (true) or no (false), i.e., the type declaration for prefix should run: `prefix :: String -> String -> Bool`.

Prefixes of a string `ys` are defined as follows:

1. `[]` is a prefix of `ys`,
1. if `xs` is a prefix of `ys` then `x:xs` is a prefix of `x:ys`,
1. nothing else is a prefix of `ys`.

Here is the code for `prefix` that implements this definition:

In [33]:
prefix :: String -> String -> Bool
prefix [] ys = True
prefix (x:xs) [] = False
prefix (x:xs) (y:ys) = (x==y) && prefix xs ys

The definition of `prefix` uses the Haskell operator `&&` for conjunction.

### Exercise 1.17

Write a function `substring :: String -> String -> Bool` that checks whether `str1` is a substring of `str2`. The substrings of an arbitrary string `ys` are given by:

1. if `xs` is a prefix of `ys`, `xs` is a substring of `ys`,
1. if `ys` equals `y:ys'` and `xs` is a substring of `ys'`, `xs` is a substring of `ys`,
1. nothing else is a substring of `ys`.

## 1.6 Haskell Types

The basic Haskell types are:

* `Int` and `Integer`, to represent integers. Elements of `Integer` are unbounded. That's why we used this type in the implementation of the prime number test.
* `Float` and `Double` represent floating point numbers. The elements of `Double` have higher precision.
* `Bool` is the type of Booleans.
* `Char` is the type of characters.

Note that the name of a type always starts with a capital letter.

To denote arbitrary types, Haskell allows the use of type *variables*. For these, `a`, `b`,
..., are used.

New types can be formed in several ways:

* By list-formation: if `a` is a type, `[a]` is the type of lists over `a`. Examples:
`[Int]` is the type of lists of integers; `[Char]` is the type of lists of characters,
or strings.
• By pair- or tuple-formation: if `a` and `b` are types, then `(a,b)` is the type of pairs with an object of type a as their first component, and an object of type `b` as their second component. Similarly, triples, quadruples, ..., can be formed. If `a`, `b` and `c` are types, then `(a,b,c)` is the type of triples with an object of type `a` as their first component, an object of type `b` as their second component, and an object of type `c` as their third component. And so on
(p. 139).
* By function definition: `a -> b` is the type of a function that takes arguments
of type `a` and returns values of type `b`.
* By defining your own data-type from scratch, with a data type declaration. More about this in due course.

Pairs will be further discussed in Section 4.5, lists and list operations in Section
4.6.

Operations are procedures for constructing objects of a certain types `b` from ingredients of a type `a`. Now such a procedure can itself be given a type: the type of a transformer from `a` type objects to `b` type objects. The type of such a procedure can be declared in Haskell as `a -> b`.

If a function takes two string arguments and returns a string then this can be
viewed as a two-stage process: the function takes a first string and returns a transformer from strings to strings. It then follows that the type is `String ->
(String -> String)`, which can be written as `String -> String -> String`, because of the Haskell convention that `->` associates to the right.

### Exercise 1.18

Find expressions with the following types:

1. `[String]`
1. `(Bool,String)`
1. `[(Bool,String)]`
1. `([Bool], String)`
1. `Bool -> Bool`

Test your answers by means of using the command `:t`.

### Exercise 1.19

Use the command `:t` to find the tpyes of the following predefined functions:

1. `head`
1. `last`
1. `init`
1. `fst`
1. `(++)`
1. `flip`
1. `flip (++)`

Next, supply these functions with arguments of the expected types, and try to guess
what these functions do.

Copyright (2004) Kees Doets and Jan van Eijk, The Haskell Road to Logic, Maths and Programming.